[우] 로또 2단계 - 수동 미션 제출합니다.

hwookim opened the pull request:
controller가 domain과 view를 연결해줬음 하면서 내용을 바꿔봤는데 잘 되었는지 모르겠네요.
아직 공부할 게 많은 것 같습니다ㅠㅠ

리뷰 잘부탁드립니다!

hwookim committed:
[Lotto] refactor : exception 패키지 분리

hwookim committed:
[Lotto] feat : LottoAmount 클래스 추가

 로또 수동/자동 구매량을 갖는 객체

 불필요한 상수 제거

hwookim committed:
[Lotto] refactor : 불필요한 상수 제거

hwookim committed:
[Lotto] docs : 수동 요구사항 추가

hwookim committed:
[Lotto] feat : InputView.inputManualLottoAmount 추가

 LottoAmount를 만들기 위한 메소드
 LottoManager 재구성

hwookim committed:
[Lotto] feat : 수동 구매 추가를 위한 Buyer 생성자 수정

 Buyer.createManualLottos 추가
 Buyer.createManualLotto 추가
 Buyer.createAutoLottos 메소드명 변경

 InputView.inputManualLottoNumbers 추가

 LottoManager 적용

hwookim committed:
[Lotto] feat : 수동 구매 추가를 위한 Buyer 생성자 수정

 Buyer.createManualLottos 추가
 Buyer.createManualLotto 추가
 Buyer.createAutoLottos 메소드명 변경

 InputView.inputManualLottoNumbers 추가

 LottoManager 적용

hwookim committed:
[Lotto] feat : 수동 구매 추가를 위한 Buyer 생성자 수정

 Buyer.createManualLottos 추가
 Buyer.createManualLotto 추가
 Buyer.createAutoLottos 메소드명 변경

 InputView.inputManualLottoNumbers 추가

 LottoManager 적용

hwookim committed:
[Lotto] feat : 최종 결과 출력 LottoManager 적용

hwookim committed:
[Lotto] fix : 프로덕션 코드 변경으로 인한 테스트코드 수정

Hue9010 commented on src/test/java/lotto/domain/BuyerTest.java:
```diff
@@ -0,0 +1,29 @@
+
+public class BuyerTest {
+    @Test
+    @DisplayName("생성 테스트")
```
지금처럼 자동, 수동이 섞인 경우라면
1. 자동만 사는 경우
2. 수동만 사는 경우
3. 자동 + 수동을 섞는 경우
각각 테스트 해주는게 좋답니다.

그랬을때 이점이 생성할 때 문제가 생겼을 때 수동/자동/수동+자동 중 어디서 문제가 생긴건지 빠르게 알 수 있는 장점이 있답니다.

hwookim replied:
테스트코드를 상황별로 분리했습니다.

분리하다보니 Buyer의 생성자에서 매개변수로 받은 manualLottos가 null이 아니면 수동 로또 구매량이 0이더라도 수동로또를 생성하는 오류가 발견되어 수정했습니다.

Hue9010 replied:
이 부분은 이미 알아채셨을 것 같은데 List가 빈 경우랍니다.

Hue9010 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -2,47 +2,55 @@
-    public LottoManager(int money) {
-        this.money = new Money(money);
-        this.buyer = new Buyer(this.money.calculateLottoTicketCount());
+    public static void run() {
```
줄 수를 줄이고자 한걸로 보이긴하지만 가독성을 위해서 개행을 일부 넣어주는 건 어떨까요? 😄

Hue9010 replied:
controller가 domain과 view를 연결해줬음 하면서 내용을 바꿔보셨다는 내용이 LottoManager일 것 같은데 그전에 확인해보고 싶은 부분이 있습니다. controller에 대해서 과정 중에 어떻게 배웠는지 궁금하네요. 아직 웹이 시작하지 않은 단계라 그에 맞춰서 알려주셨을 수도 있을 것 같아서 현재 학습하신 방향에 맞춰서 안내를 드리고 싶어서 문의드립니다!

hwookim replied:
![image](https://user-images.githubusercontent.com/45786387/75604469-34b96a80-5b1c-11ea-9320-4416d3e892f6.png)

과정 중에는 이 그림을 예시로 들었습니다.
디자인 패턴에 대해서 자세히 언급하진 않았던 걸로 기억하는데... 
제가 그냥 이곳저곳 주워 들으며 공부해서 맞는지는 잘 모르겠네요ㅠㅠ

일단은 '도메인과 뷰 사이에서 값을 전달해주고 관리하는 것'으로 이해하고 있습니다.

2단계에서 LottoManager를 수정 한 이유는 프로그램의 흐름을 한눈에 보고 싶어서 입니다.

1단계 코드에서는 main함수에서 view를 통해 입력값을 받은 뒤,
LottoManager에서 domain의 결과값을 다시 main으로 보내 출력을 하는 방식이었습니다.
LottoManager와 main이 함께 컨트롤러였죠.

그러다보니 다시 수정하려고 프로그램의 흐름을 확인할 때 불편했습니다.
main에서만 보면 중간 과정이 모두 끊겨 보이고 두 클래스를 왔다갔다하며 흐름을 확인하니 헷갈릴 수 있겠다는 느낌이 들었습니다.
그래서 하나의 클래스 내에서 view와 domain을 연결해주면 흐름을 한눈에 파악할 수 있지 않을까 하여 이렇게 변경했습니다.

다시 생각해보니 지금은 프로그램이 작아 하나의 클래스에 어찌저찌 우겨넣었지만
로직이 복잡해지면 이게 더 나쁜 습관인 것 같네요.
지금도 run 메소드가 너무 길어진 감이 있고요...ㅠㅠ

Hue9010 replied:
```java
public class RacingMain {
    public static void main(String[] args) {
        String carNames = InputView.getCarNames();
        int tryNo = InputView.getTryNo();

        RacingGame racingGame = new RacingGame(carNames, tryNo);
        while(!racingGame.isEnd()) {
            racingGame.race();
            ResultView.printCars(racingGame.getCars());
        }
        ResultView.printWinners(racingGame.getWinners());
    }
}
```
어떻게 바꿀지 가이드를 드리고 싶어서 좀 고민하다가 늦어지고 있네요. 그래서 일단 가이드는 별개로 먼저 답변드리는게 나을 것 같아서 코멘트답니다! 이 부분을 처음부터 고려했어야 했는데 지금 학습 단계는 아니라고 판단해서 미리 챙기지 못했네요. 😭 

그림을 보면 Controller에서 View를 통해 값을 입력받는게 아니라 View에서 입력 받은 값을 Domain에 전달해주는 역할을 합니다. 그리고 Domain을 통해 받은 실행 결과를 다시 view에 전달합니다.

여기서 핵심은 
`controller가 domain과 view를 연결해줬음 하면서` + `도메인과 뷰 사이에서 값을 전달해주고 관리하는 것` -> `domain과 view가 controller를 사이에 두고 분리 되어 있다` 가 됩니다.

잠깐 이 부분은 저도 조금 더 정리하고 초기 가이드를 잡아드리는게 좋을 것 같네요.

Hue9010 replied:
어제 몸상태가 좋지 못해서 하루 종일 누워있다 답변을 못드렸네요.. 😭 
이따 저녁때 좀더 상세하게 달아드릴 계획인데 그전에 다시한번 간단히 말씀드리면 위 예시로 있는 코드는 한번의 입력/출력만 존재하는 상태입니다. (차이름, 횟수만 입력하면 추가 입력 없이 도메인에 연결하고 출력을 해줍니다.)

입력 View
```java
String carNames = InputView.getCarNames();
        int tryNo = InputView.getTryNo();
```

모델
```java
        RacingGame racingGame = new RacingGame(carNames, tryNo);
        while(!racingGame.isEnd()) {
            racingGame.race();
            ResultView.printCars(racingGame.getCars());
        }
```

출력 View
```java
        while(!racingGame.isEnd()) {
            racingGame.race();
            ResultView.printCars(racingGame.getCars());
        }
        ResultView.printWinners(racingGame.getWinners());
```

이번 로또 같은 경우는 여러번의 입력 단계와 그에 따른 출력단계가 존재합니다.
크게 아래와 같이 나눌 수 있을 것 같네요.(입력 값 부분은 수동 입력 부분이 조금 나뉠 수도 있겠지만)
금액, 횟수, 수동 횟수, 수동 번호 입력 -> 입력 값 출력
당첨 번호 -> 당첨 결과 출력

View를 통해서 입력 받은 값을 도메인에 전달하고 그에 따른 결과를 출력 View에 전달해준다는 걸 고려해서 의미상으로만 나눠서 run 메서드 내부를 한번 구성해보세요!
바로 domain에 인자값으로 넘겨 주는 것도 좋지만 구분이 눈에 명확하게되게 Racing 예제처럼 입력값을 직접 받은 다음에 넘겨주게 해보세요!

hwookim replied:
말씀하신 의도가 맞는지 자신이 없지만....😭  

금액 등 입력 ~ 로또 번호 출력 : purchaseLotto()
당첨 번호 입력 ~ 당첨 결과 출력 : analyzeLotto()
으로 나누었습니다.

두 단계에서 주고 받는 값이 buyer와 money로 두가지라 필드 변수를 추가했습니다.

> View를 통해서 입력 받은 값을 도메인에 전달하고 그에 따른 결과를 출력 View에 전달해준다는 걸 고려해서 의미상으로만 나눠서 run 메서드 내부를 한번 구성해보세요!

이 부분에 대한 작업인데 아직도 약간 아리송합니다 ㅠㅠ
제가 제대로 이해하고 적용한게 맞을까요? 😢

Hue9010 replied:
이 부분은 어떻게 풀어서 설명 드릴지 많이 고민하다가 늦어지기도 하고 설명도 매끄럽진 못했네요. 아마 인터넷을 찾아봐도 명확하지도 않을 이유 중 하나가 자바로 찾으시면 주로 스프링을 통한 설명이다 보니 지금과 같은 콘솔 자바 애플리케이션에 접목해서 생각하기가 쉽진 않을 겁니다. 형태도 생각보다 다르고요.

사실 처음부터 잘 작성하셨답니다. 다만 형태가 제대로 눈에 띄지 않아서 인풋의 값들을 빼보고 입력, 도메인 연결, 출력의 흐름에 맞춰서 다시한번 해보길 제안 드렸던겁니다.

같이 좀더 얘기해보면서 차근 차근 알아가보고 싶은 부분도 있지만 온라인이라는 제약도 있고 그러다보니 의미 전달을 제대로 못해드려서 불필요한 고민의 시간을 드릴 것 같아 예제 중심으로 한번 보면

```java
public class LottoManager {
    public static void run() {
        // InputView
        int budget = InputView.inputMoney();
        int manualAmount = InputView.inputManualLottoAmount();

        // Domain
        Money money = new Money(budget);
        int totalAmount = money.calculateTotalLottoAmount();
        LottoAmount lottoAmount = new LottoAmount(totalAmount, manualAmount);

        // OutPutView
        OutputView.printLottoAmount(lottoAmount);

        ///////////////////////////////////////////////////////////////////////////////////////////////
        ///////////////////////////////////////////////////////////////////////////////////////////////

        // InputView
        List<String> manualLottoNumbers = InputView.inputManualLottoNumbers(lottoAmount.getManualLottoAmount());

        // Domain
        Buyer buyer = new Buyer(manualLottoNumbers, lottoAmount);

        // OutPutView
        OutputView.printLottoNumbers(buyer.getLottos());

        ///////////////////////////////////////////////////////////////////////////////////////////////
        ///////////////////////////////////////////////////////////////////////////////////////////////

        // InputView
        String input = InputView.inputWinningLottoNumbers();
        int bonusNumber = InputView.inputBonusNumber();

        List<LottoNumber> winningLottoNumbers = ConvertInput.convertLottoNumbers(input);

        // Domain
        WinningLotto winningLotto = new WinningLotto(winningLottoNumbers, bonusNumber);
        LottoResult lottoResult = new LottoResult(buyer, winningLotto);

        // OutPutView
        OutputView.printLottoResults(lottoResult.getLottoResult());
        OutputView.printRewardRate(lottoResult.calculateRewardRate(money.getMoney()));
    }
}
```
(의도적으로 InputView -> Domain -> OutputView 흐름이 부각되게 `OutputView.printLottoAmount(lottoAmount);` 의 위치를 변경했습니다.)

보면 InputView를 통해 받아온 입력 값을 도메인에 전달 해주고, 도메인에 의한 실행결과 값을(생성의 결과도 포함) OutPutView에 전달해주는게 눈에 보이실겁니다. LottoManager라는 컨트롤러가 View와 Domain을 연결해주고 있죠? 😄

MVC가 아니였다면? 아주 간단한 예로 Model과 View과 Controller 없이 직접 연결된 경우가 있겠네요.
```java
public class Money {
    private static final int LOTTO_PRICE = 1000;

    private int money;

    public Money() {
        int inputMoney = InputView.inputMoney();
        validateMoney(inputMoney);
        this.money = inputMoney;
    }
```

혹시 아리쏭한 부분 있으면 더 여쭤봐주세요!!

hwookim replied:
휴의 코드가 `view와 domain을 잇는 controller`에는 더 부합해 보이는군요!
같은 코드라도 보여주고 싶은 중점에 따라 표현 방식이 달라지니.. 생각할게 많은 것 같네요 😭 
친절한 답변 감사드립니다! 👍 

앞선 얘기들과는 좀 별개인 질문입니다.
미션에 각 메소드를 10라인 이하로 작성하라는 제한 사항이 있어서 생긴 고민이에요.

run 메소드를 분리하게 될 때, 어떤 방식으로 분리해야 코드가 조금 더 깔끔할까요?

```java
public class LottoManager {
    public static void run() {
        Money money = readMoney();
        LottoAmount lottoAmount = calculateLottoAmount(money);
        OutputView.printLottoAmount(lottoAmount);

        Buyer buyer = readLottos(lottoAmount);
        OutputView.printLottoNumbers(buyer.getLottos());

        LottoResult lottoResult = analyzeLotto(buyer);
        OutputView.printLottoResults(lottoResult.getLottoResult());
        OutputView.printRewardRate(lottoResult.calculateRewardRate(money.getMoney()));
    }

    private static Money readMoney() {
        int budget = InputView.inputMoney();

        return new Money(budget);
    }

    private static LottoAmount calculateLottoAmount(Money money) {
        int manualAmount = InputView.inputManualLottoAmount();
        int totalAmount = money.calculateTotalLottoAmount();

        return new LottoAmount(totalAmount, manualAmount);
    }

    private static Buyer readLottos(LottoAmount lottoAmount) {
        List<String> manualLottoNumbers
                = InputView.inputManualLottoNumbers(lottoAmount.getManualLottoAmount());

        return new Buyer(manualLottoNumbers, lottoAmount);
    }

    private static LottoResult analyzeLotto(Buyer buyer) {
        WinningLotto winningLotto = readWinningLotto();

        return new LottoResult(buyer, winningLotto);
    }

    private static WinningLotto readWinningLotto() {
        String input = InputView.inputWinningLottoNumbers();
        List<LottoNumber> winningLottoNumbers = ConvertInput.convertLottoNumbers(input);
        int bonusNumber = InputView.inputBonusNumber();

        return new WinningLotto(winningLottoNumbers, bonusNumber);
    }
}
```
'10줄' 이라는 제한에 얽매여서 억지로 분리한 느낌이 들고,
이렇게 되면 휴가 말해주셨던 view - domain - view 의 흐름도 완전히 깨져버리네요ㅠㅠ

휴라면 어떤 식으로 메소드 분리를 진행하실지 궁금합니다 😢

Hue9010 replied:
아니요 이번 제 코드도 우가 작성했던 코드도 이미 컨트롤러의 역할을 제대로 수행하고 있다는 걸 보여주기 위해서 view - domain - view의 흐름을 억지로 부각되게 작성했을 뿐입니다.

```java
// 1번
    private static Money readMoney() {
        int budget = InputView.inputMoney();

        return new Money(budget);
    }

// 2번
    private static Money readMoney() {
        return new Money(InputView.inputMoney());
    }
```
대표적인게 입력 값을 원시타입으로 받아 도메인에 전달해주는 부분인데 입력 받은 값을 도메인한테 전달해 준다는 걸 부각 해주기 위해 1번과 같이 짰을 뿐이지 기존 코드였던 2번 처럼 작성 한 것도 실제론 컨트롤러가 입력 값을 도메인에 연결해 주는 거랍니다.

제약 사항 중 콘솔 화면상의 입출력이 이미 정해져 있다는 겁니다. 그러다보니 정해진 형태의 입출력에 맞추다보니 힘든것도 있답니다.

저라면 어떻할지 부분에 대해서는 run()메서드에서 개행으로 표시 해준 세개로 나누거나 위의 두개는 합쳐서 나눌 것 같네요.

그리고 readMoney()를 따로 빼지 않고 new Money(InputView.inputMoney())로 둘테고요. 어디까지나 View를 통해 값을 받아다가 도메인에 연결해준 다는 걸 부각하기 위해서 제안드렸던 부분입니다.

지금보다보니 money.calculateTotalLottoAmount()가 아니라 총 횟수도 LottoAmount가 돈을 받으면 계산하게 옮길 수도 있었겠네요. 그렇다면 
```java
    private static LottoAmount calculateLottoAmount(Money money) {
        int manualAmount = InputView.inputManualLottoAmount();
        int totalAmount = money.calculateTotalLottoAmount();

        return new LottoAmount(totalAmount, manualAmount);
    }
```
이 부분도 단순해질 수 있겠네요.

```java
    private static WinningLotto readWinningLotto() {
        String input = InputView.inputWinningLottoNumbers();
        List<LottoNumber> winningLottoNumbers = ConvertInput.convertLottoNumbers(input);
        int bonusNumber = InputView.inputBonusNumber();

        return new WinningLotto(winningLottoNumbers, bonusNumber);
    }
```
이 부분은 WinningLotto 혹은 WinningLotto를 만들어 줄 클래스를 만든 후 input, bonusNumber를 가지고 WinningLottot를 만들어줄 메서드를 만들어 줄 것 같습니다. 혹은 LottoNumber에서 문자열받으면 그걸 기반으로  List<LottoNumber>를 만들어줄 수도 있겠네요.

이와같이 하려는 이유는 컨트롤러는 View와 도메인을 연결하는 것에만 집중하고 그외에는 가능하면 최소한의 로직도 포함하지 않으려고 하기 위함입니다.  😄 

혹시나 싶어 다시 말씀드리는데
`휴가 말해주셨던 view - domain - view 의 흐름` 제가 짰던 코드는 어디까지나 그 흐름이 눈에 더 잘  들어나게 하기 위해서 짠 코드 일 뿐 기존 우의 코드 또한 컨트롤러로서 역할은 수행하고 있었습니다. 👍

Hue9010 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -2,47 +2,55 @@
-        lottoResult = new LottoResult();
-        lottoResult.calculateLottoResult(buyer.getLottos(), winningLotto);
-        return lottoResult.getLottoResult();
+    private static Buyer makeLottos(LottoAmount lottoAmount) {
```
로또의 생성은 바이어가 담당했던걸로 보이는데 그 생성의 책임이 일부 LottoManager로 넘어 온걸로 보이네요.
LottoFactory로 바꾸는 것도 생각해 볼 수 있을 것 같습니다. 이후 일반 Lotto와 WinningLotto에 대한 생성 책임까지 부여해줄수 있겠네요.

Hue9010 replied:
analyzeLotto, analyzeRwardRate는 따로 메서드로 뺄 필요가 있을까요?

hwookim replied:
로또의 생성은 여전히 Buyer에서 관리하고 있습니다.
LottoManager의 run 메소드를 좀 더 간결하게 하기 위해서, 프로그램의 흐름에 대한 가독성을 높이기 위해서 makeLottos 라는 별도의 메소드를 만들었습니다.
makeLottos 메소드는 InputView에서 수동 로또 번호를 받아 Buyer에 넘겨줄 뿐 실제 생성은 Buyer 내에서 진행됩니다.

초기에 설계할 때 로또 여러장을 가지고 있는건 구매자라고 생각하여 Buyer라는 이름으로 클래스명을 지었습니다.
그렇기에 WinningLotto는 밖에서 정의되고 컨트롤러에 의해 값 비교가 이뤄지는 식으로 진행하였습니다.
WinningLotto에 대한 생성 책임까지 Buyer가 지게 된다면 따로 필드변수로 WinningLotto를 가지게 하라는 말씀이실까요?

hwookim replied:
analyzeLotto, analyzeRewardRate는 run 메소드 내에서 어떤 동작을 하는 부분인지 명시적으로 보여주려는 목적이었는데, 따로 분리하지 않아도 충분히 그 의도를 보여주는 것 같네요.
메소드를 삭제했습니다. 😃

Hue9010 replied:
makeLottos의 경우는 확실히 말씀 주신 것처럼 Buyer에게 입력값을 전달해주기만 하는데 생성되는건 Buyer인데 메서드 명은 makeLottos라서 의아한 부분이 있었습니다. 😅 
readWinningLotto의 경우는 입력 값에 대하여 ConvertInput을 통해 List<LottoNumber>로 변환 후 WinningLotto에 전달해주는 부분이 존재하여 말씀드렸네요. 

```java
    private static Buyer makeLottos(LottoAmount lottoAmount) {
        List<String> manualLottos =
                InputView.inputManualLottoNumbers(lottoAmount.getManualLottoAmount());
        return new Buyer(manualLottos, lottoAmount);
    }

    private static WinningLotto readWinningLotto() {
        List<LottoNumber> winningLottoNumbers
                = ConvertInput.convertLottoNumbers(InputView.inputWinningLottoNumbers());
        int bonusNumber = InputView.inputBonusNumber();

        return new WinningLotto(winningLottoNumbers, bonusNumber);
    }
```

사실 Buyer를 구매자로 사용하신 건지 중개상과 같은 의미로 사용하신건지 확신이 안섰네요. 객체지향의 세계는 현실과 같을 필요는 없어서 구매자가 자신의 로또를 만드는 것도 충분히 가능하니 변경은 안하셔도 될 것 같습니다!

만약 약간의 변경을 해본다면 아래와 같이도 가능하겠네요
- LottoFactory를 추가 구현하여 Lotto와 WinningLotto 생성(둘 다 Lotto 타입이므로 LottoFactory가 둘 다 만들어줘도 이상 x) -> 그 후 Buyer한테 전달
- Buyer(구매자)가 자신의 Lotto와 WinningLotto를 비교하여 그 결과에 따라 LottoResult 생성

아주 혹시라도 변경은 하지마시고요! 어디까지나 다양한 방법이 있다는 예시일 뿐이니까요 😋

Hue9010 commented on src/main/java/lotto/domain/Buyer.java:
```diff
@@ -1,19 +1,35 @@
 
-    private void createLottos(int lottoTicketCount) {
+    private void createManualLottos(List<String> manualLottos) {
+        if (null == manualLottos) return;
```
`manualLottos != manualLottos` 도 가능하겠네요

hwookim replied:
Buyer의 생성자를 수정하면서 해당 부분을 다른 방식으로 검사하도록 변경했습니다.

```java
public Buyer(List<String> manualLottos, LottoAmount lottoAmount) {
    if (lottoAmount.getManualLottoAmount() != 0) {
        createManualLottos(manualLottos);
    }
    createAutoLottos(lottoAmount.getAutoLottoAmount());
}
```

hwookim replied:
HUE가 말해주신 코드가 어떤식으로 작동되는건지 잘 모르겠습니다.
manualLottos가 null이라는 가정 하에 해당 비교가 이루어지면 NPE가 발생하지는 않나요?

Hue9010 replied:
앗 오타가 있었네요 🙀 
`manualLottos != null`을 의도했었던 겁니다. 😨 
1. null일땐 return, null이 아니면 lottos.addAll해라
2. null이 아니면 lottos.addAll해라

기존 코드같은경우는 1번처럼 두가지를 나타내지만 2번 과정이 더 단순하게 생각할 수 있지 않을가 생각되어 제안드렸습니다.

Hue9010 reviewed (changes requested):
수정 로또 추가를 다른 곳에 변경이 적은 상태로 잘해주셨네요! 💯 
변경이 적은 상태로 기능을 추가하셨는데 기존 테스트 코드들이 많이 도움이 되던가요? 이 부분도 어떠셨는지 궁금하네요 😸 
몇가지 피드백과 문의주신 사항에대해서는 피드백 드리기전에 한가지 질문드렸는데 확인 부탁드려요. 😄

hwookim committed:
[Lotto] refactor : 테스트코드 수정

 DisplayName 추가
 테스트 케이스 상황별 분리

hwookim committed:
[Lotto] fix : Buyer 생성 오류 수정

 LottoAmount.manualLottoAmount가 0이더라도 매개변수 manualLottos가 null이 아니면 생성하는 문제 수정
 테스트코드 상황 별 분리를 통한 생성 오류 해결 확인

hwookim committed:
[Lotto] refactor : 불필요한 메소드 삭제

 analyzeLotto(), analyzeRewardRate() 삭제, run()에 통합

hwookim committed:
[Lotto] refactor : 메소드 분리

 Money, Buyer 필드변수 추가
 역할에 따른 run 메소드 분리

Hue9010 reviewed (approved):
요구사항에 맞춰서는 이미 잘 구현해주신데다 이전 피드백 반영도 잘하셔서 이번 단계는 여기서 머지하도록 할게요 👍 대신 생각해보고 질답하는 시간은 필요할 것 같네요😸 
코멘트로 답변 드린 부분 확인하시고 코멘트나 DM으로 편히 문의주세요!
